tcp.c

```diff
@@ -537,14 +542,13 @@ tcp_connect(char *server)
 void
 tcp_disconnect(void)
 {
-	if (g_ssl)
-	{
-		if (!g_network_error)
-			(void) SSL_shutdown(g_ssl);
-		SSL_free(g_ssl);
-		g_ssl = NULL;
-		SSL_CTX_free(g_ssl_ctx);
-		g_ssl_ctx = NULL;
+	if (g_ssl_initialized) {
+		//gnutls_bye(g_tls_session, GNUTLS_SHUT_RDWR);
```

Comment:
Why is this commented out?

Reply:
Please have another look. It seems very wrong to not do a proper shutdown. If you get stuck then at least include a FIXME comment that this should be investigated further.

Reply:
SUre. I'll do it because I don't like this approach either.

Reply:
Basically there was an unfortunate combination of two bugs:
1) I'd better use ( it's way faster)
````C
gnutls_bye(g_tls_session, GNUTLS_SHUT_WR)
````
 instead of
````C
gnutls_bye(g_tls_session, GNUTLS_SHUT_RDWR)
````
2) I zeroed rcvd in **tcp_recv**() forming a perfect infinite loop:)
Check **tcp.c: 240-245**